src/nlp_utils.jl

```diff
@@ -23,3 +25,39 @@ macro rangecheck(lo, hi, vars...)
   end
   Expr(:block, exprs...)
 end
+
+"""
+    coo_prod!(rows, cols, vals, v, Av)
+
+Compute the product of a matrix `A` given by `(rows, cols, vals)` and the vector `v`.
+The result is stored in `Av`, which should have length equals to the number of rows of `A`.
+"""
+function coo_prod!(rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer}, vals :: AbstractVector, v :: AbstractVector, Av :: AbstractVector)
+  fill!(Av, zero(eltype(v)))
+  nnz = length(rows)
+  @inbounds for k = 1:nnz
+    i, j = rows[k], cols[k]
+    Av[i] += vals[k] * v[j]
+  end
+  return Av
+end
+
+"""
+    coo_sym_prod!(rows, cols, vals, v, Av)
+
+Compute the product of a symmetric matrix `A` given by `(rows, cols, vals)` and the vector `v`.
+The result is stored in `Av`, which should have length equals to the number of rows of `A`.
+Only half triangle should be passed.
```

Comment:
Only one triangle of `A` should be passed.

Reply:
Fixed